test/e2e/apimachinery/aggregated_discovery.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apiextensions-apiserver/test/integration/fixtures"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/names"
+	clientdiscovery "k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubernetes/test/e2e/framework"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+)
+
+var _ = SIGDescribe("AggregatedDiscovery", func() {
+	f := framework.NewDefaultFramework("aggregateddiscovery")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	/*
+		Release : v1.30
+		Testname: Aggregated Discovery Interface
+		Description: An apiserver MUST support the Aggregated Discovery client interface. Built-in resources MUST all be present.
+	*/
+
+	ginkgo.It("should support aggregated discovery interface", func(ctx context.Context) {
+		d := f.ClientSet.Discovery()
+
+		ad, ok := d.(clientdiscovery.AggregatedDiscoveryInterface)
+		if !ok {
+			framework.Failf("Expected client to support aggregated discovery")
+		}
+		serverGroups, resourcesByGV, _, err := ad.GroupsAndMaybeResources()
```

Comment:
I'm pretty sure this function falls back to old discovery.  Perhaps you want to use a raw client to issue GETs?

Reply:
This is specifically part of the aggregated discovery interface https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/discovery_client.go#L92-L96

Yes, `serverGroups` may be populated via the old discovery if aggregated discovery is missing, but `resourcesByGV` will be empty as this only issues one request to the server. The test is verifying that resourcesByGV is populated correctly which is an aggregated discovery only feature.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/discovery_client.go#L275-L287

I have a slight preference for using the client for testing since that's how client-go and other controllers will be interacting with the feature.

Reply:
Could we test with client-go AND test with a raw client? This does seem like a case where e2e testing the wire protocol is worthwhile.

Reply:
Ack, added tests for the raw client